src/main/kotlin/com/sourcegraph/cody/edit/FixupSession.kt

```diff
@@ -0,0 +1,354 @@
+  private val logger = Logger.getInstance(FixupSession::class.java)
+  val project = editor.project!!
+
+  // This is passed back by the Agent when we initiate the editing task.
+  var taskId: String? = null
+
+  var performedEdits = false
+    private set
+
+  private var lensGroup: LensWidgetGroup? = null
+
+  private var selectionRange: Range? = null
+
+  private var rangeMarkers: MutableSet<RangeMarker> = mutableSetOf()
+
+  private val lensActionCallbacks =
+      mapOf(
+          COMMAND_ACCEPT to { accept() },
+          COMMAND_CANCEL to { cancel() },
+          COMMAND_RETRY to { retry() },
+          COMMAND_DIFF to { diff() },
+          COMMAND_UNDO to { undo() },
+      )
+
+  init {
+    triggerDocumentCodeAsync()
+  }
+
+  fun commandCallbacks(): Map<String, () -> Unit> = lensActionCallbacks
+
+  @RequiresEdt
+  private fun triggerDocumentCodeAsync() {
+    // This caret lookup requires us to be on the EDT.
+    val caret = editor.caretModel.primaryCaret.offset
+
+    FixupService.backgroundThread {
+      val project = editor.project!!
+      CodyAgentService.withAgent(project) { agent ->
+        workAroundUninitializedCodebase(editor)
+        // Force a round-trip to get folding ranges before showing lenses.
+        ensureSelectionRange(agent, editor, caret)
+        showWorkingGroup()
+        // All this because we can get the workspace/edit before the request returns!
+        FixupService.getInstance(project).addSession(this) // puts in Pending
+        makeEditingRequest(agent)
+            .handle { result, error ->
+              if (error != null || result == null) {
+                // TODO: Adapt logic from CodyCompletionsManager.handleError
+                logger.warn("Error while generating doc string: $error")
+                FixupService.getInstance(project).removeSession(this)
+              } else {
+                taskId = result.id
+                selectionRange = result.selectionRange
+                FixupService.getInstance(project).addSession(this)
+              }
+              null
+            }
+            .exceptionally { error: Throwable? ->
+              if (!(error is CancellationException || error is CompletionException)) {
+                logger.warn("Error while generating doc string: $error")
+              }
+              FixupService.getInstance(project).removeSession(this)
+              null
+            }
+            .completeOnTimeout(null, 3, TimeUnit.SECONDS)
+      }
+    }
+  }
+
+  // We're consistently triggering the 'retrieved codebase context before initialization' error
+  // in ContextProvider.ts. It's a different initialization path from completions & chat.
+  // Calling onFileOpened forces the right initialization path.
+  private fun workAroundUninitializedCodebase(editor: Editor) {
+    val file = FileDocumentManager.getInstance().getFile(editor.document)!!
+    val project = editor.project!!
```

Comment:
Please let's do not rely on `!!`, it's not always the case that project is present.

Reply:
Fixed.